tfhe/src/integer/block_decomposition.rs

```diff
@@ -92,27 +94,35 @@ pub struct BlockDecomposer<T> {
     bit_mask: T,
     num_bits_in_mask: u32,
     num_bits_valid: u32,
-    padding_bit: T,
+    // This value (0 or 1) will be used to pad the last block
+    // when the number of bits in the block is not a multiple of T::BITS
+    padding_bit: Option<u32>,
     limit: Option<T>,
 }
 
 impl<T> BlockDecomposer<T>
 where
     T: Decomposable,
 {
+    /// Creates a block decomposer that will stop when the value reaches zero
     pub fn with_early_stop_at_zero(value: T, bits_per_block: u32) -> Self {
-        Self::new_(value, bits_per_block, Some(T::ZERO), T::ZERO)
+        Self::new_(value, bits_per_block, Some(T::ZERO), None)
     }
 
-    pub fn with_padding_bit(value: T, bits_per_block: u32, padding_bit: T) -> Self {
-        Self::new_(value, bits_per_block, None, padding_bit)
+    /// Creates a block decomposer that will set the surplus bits to a specific value
+    /// when bits_per_block is not a multiple of T::BITS
+    pub fn with_padding_bit(value: T, bits_per_block: u32, padding_bit: u32) -> Self {
+        Self::new_(value, bits_per_block, None, Some(padding_bit))
     }
 
     pub fn new(value: T, bits_per_block: u32) -> Self {
-        Self::new_(value, bits_per_block, None, T::ZERO)
+        Self::new_(value, bits_per_block, None, None)
     }
 
-    fn new_(value: T, bits_per_block: u32, limit: Option<T>, padding_bit: T) -> Self {
+    fn new_(value: T, bits_per_block: u32, limit: Option<T>, padding_bit: Option<u32>) -> Self {
+        if let Some(padding_bit) = padding_bit {
+            assert!(padding_bit <= 1);
```

Comment:
how about a PaddingBit struct which holds an Option<?> with repr(transparent) and a new_one(), new_zero() function ?

? can be bool or u32 as you like

will avoid the assert to enforce the invariant, wdyt ?

you could even have consts like PADDING_BIT_ZERO, PADDING_BIT_ONE if you like

Reply:
I think bool would be enough

Reply:
ah, don't you need the Option thing to discrimnate some specific case ?

Reply:
Its only needed in the struct itself and in the private (new_) constructor as we use a different constructor if the user whishes to specify the padding

Reply:
I just don't know if the bool is the best tool to express what the user wants, don't recall if this struct is pub or pub(crate)

Reply:
The struct is pub

Reply:
Then I don't feel like bool is the best to carry intent here

Reply:
So you'd prefer a enum `PaddingBitValue { Zero, One }` and possibly None ?

Reply:
Don’t you think ?

Reply:
integer is fine to break